Run flynt to convert %-formatting to f-strings

eigenfoo opened the pull request:
~[WIP, will request review when ready]~

A follow-up to #139 - I simply used the [`flynt` CLI tool](https://github.com/ikamensh/flynt) to automatically convert %-formatted strings to f-strings. However, not all formatted strings have been converted to f-strings - some strings are tricky to convert, or `flynt` simply fails on them - in which case I've just left the string as-is.

It's important to note that %-formatting are different from f-strings [in meaningful ways](https://github.com/ikamensh/flynt#dangers-of-conversion), so it's important to check the tests.

---

For posterity, here's the output of `flynt`:

```
Running flynt v.0.56

Flynt run has finished. Stats:

Execution time:                            19.170s
Files modified:                            125
Character count reduction:                 2509 (0.03%)

Per expression type:
Old style (`%`) expressions attempted:     846/1483 (57.0%)
`.format(...)` calls attempted:            286/291 (98.3%)
No concatenations attempted.
F-string expressions created:              799
Out of all attempted transforms, 321 resulted in errors.
To find out specific error messages, use --verbose flag.

_-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_._-_.
Please run your tests before committing. Did flynt get a perfect conversion? give it a star at:
~ https://github.com/ikamensh/flynt ~
Thank you for using flynt. Upgrade more projects and recommend it to your colleagues!
```

brandonwillard reviewed (changes requested):
Changes needed:
- [x] I believe that `theano/_version.py` is generated automatically by `versioneer`; if so, we're better off regenerating that file and/or leaving it untouched.
- [x] There are _a lot_ of f-strings containing superfluous uses of `str`, `dict`, and `locals`.  

It would seem as though `flynt` isn't aware of these redundancies.  If you're interested in writing your own refactoring code that _is_ aware of such things, check out [LibCST](https://github.com/Instagram/LibCST)!

eigenfoo commented:
@brandonwillard not to dismiss your suggestion, but I just didn't feel like adding to my cognitive load by learning a new tool - I just `rg`ed the diff to find offending instances of `locals()`, `dict`s and `str`s and manually fixed them with some help from Vim macros.

twiecki commented:
@eigenfoo seems like tests still fail.

eigenfoo commented:
The only test that's failing is the patch coverage, which seems to be failing for all PRs that don't test 100% of their changes. Since this PR doesn't change any actual code behavior, I think that's okay.

twiecki commented:
Gotcha, cool.

On Fri, Nov 6, 2020, 16:22 George Ho <notifications@github.com> wrote:

> The only test that's failing is the patch coverage, which seems to be
> failing for all PRs that don't test 100% of their changes. Since this PR
> doesn't change any actual code behavior, I think that's okay.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/Theano-PyMC/pull/144#issuecomment-723137467>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFETGHUTRVWLGBZMNXJPOLSOQIB5ANCNFSM4TJPPYAQ>
> .
>

brandonwillard commented:
> ...I just didn't feel like adding to my cognitive load by learning a new tool...

Sorry, I wasn't actually suggesting that you use LibCST for the work in this PR.  I was pointing you in the direction of a library that's quite useful for creating your own fully-automated Python refactorings.  It's more of a long-term suggestion, since I strongly believe that Python needs more/better refactoring tools and libraries and I think that's more likely to happen if/when libraries like LibCST are given more attention.

brandonwillard commented on theano/compile/profiling.py:
```diff
@@ -1452,7 +1452,7 @@ def print_stats(stats1, stats2):
                 size = "%10s" % "Unknown"
```
I wonder why it didn't pick up this one.

brandonwillard commented on theano/gof/cc.py:
```diff
@@ -888,7 +885,7 @@ def code_gen(self):
-                raise NotImplementedError("%s cannot produce C code" % op)
+                raise NotImplementedError(f"{op} cannot produce C code")
             assert isinstance(behavior, str), (
                 str(node.op) + " didn't return a string for c_code"
```
This looks like a good place for an f-string.

brandonwillard commented on theano/gof/compiledir.py:
```diff
@@ -173,7 +173,7 @@ def print_compiledir_content():
+    print_title(f"Theano cache: {compiledir}", overline="=", underline="=")
     print()
 
     print_title("List of %d compiled individual ops" % len(table), underline="+")
```
This could also be an f-string.

brandonwillard commented on theano/gpuarray/sort.py:
```diff
@@ -224,8 +223,8 @@ def c_code(self, node, nodename, inps, outs, sub):
         dims = "".join("dims[%d], " % i for i in reordered_axes[1:])
```
This could be an f-string, too.

brandonwillard commented on theano/gpuarray/sort.py:
```diff
@@ -240,8 +239,8 @@ def c_code(self, node, nodename, inps, outs, sub):
-            params_di = "{}->ga.data, {}->ga.offset,\n".format(yi, yi)
+            def_distrides = f"const ssize_t *distrides = PyGpuArray_STRIDES({yi})"
+            params_di = f"{yi}->ga.data, {yi}->ga.offset,\n"
             params_di += "".join("distrides[%d], " % i for i in reordered_axes)
```
Same here.

brandonwillard commented on theano/gpuarray/type.py:
```diff
@@ -235,9 +233,7 @@ def __repr__(self):
                 bcast = str(b)
             else:
                 bcast = "%iD" % len(b)
```
This could also be an f-string.

brandonwillard commented on theano/gradient.py:
```diff
@@ -838,9 +837,9 @@ def _node_to_pattern(node):
                 raise TypeError(
-                    "%s.connection_pattern should return" % node.op
+                    f"{node.op}.connection_pattern should return"
                     + " a list of lists, but element %d" % ii
```
Another f-string candidate.

brandonwillard commented on theano/misc/check_blas.py:
```diff
@@ -293,7 +293,7 @@ def test():
         print("(%d, %d) and (%d, %d)." % (M, N, N, K))
```
This could also be an f-string.

brandonwillard commented on theano/misc/hooks/check_whitespace.py:
```diff
@@ -45,9 +45,9 @@ def get_parse_error(code):
-        return "Indentation error: %s" % err
+        return f"Indentation error: {err}"
     except tabnanny.NannyNag as err:
         return "Ambiguous tab at line %d; line is '%s'." % (
```
Another f-string candidate.

brandonwillard commented on theano/printing.py:
```diff
@@ -519,7 +519,7 @@ def process(self, output, pstate):
                 " not the result of an operation" % self.function
```
Another f-string candidate.

brandonwillard reviewed (changes requested):
Almost there!

It looks like Flynt is restrictively shy about
-  concatenated strings (e.g. `"blah %s " "blah" % "blah"`) and
- interpolations that specify integer presentations (e.g. `"%d"`).

The former can be handled by surrounding the concatenated strings with parentheses and the latter by casting to `int` in the f-string expression.

Otherwise, if I recall, there are some places where the old `%` interpolation is used to create string templates (e.g. `st = "blah %d"; st % 5`).  I don't know if f-strings can be used in that way, so we might need to convert those to `str.format` or `string.Template` forms.  We can create separate issue for that, though.

eigenfoo commented:
I've removed most instances of `%d` - I've replaced them with `f"...{int(x)}..."` in all places, unless it was obvious that it would be an integer instead of a float (e.g. `f"...{len(x)}..."`). I left them alone when they were in templated strings, which should probably be dealt with in a separate issue/PR.

eigenfoo commented:
Closed in favor of https://github.com/pymc-devs/Theano-PyMC/pull/162 - we now fork and branch, instead of working off of branches in this repo.
